### PR TITLE
Sort class members by privacy class and then by name

### PR DESCRIPTION
**Pull request by ntamas**
This PR changes the class page template in the HTML output to sort class members by privacy class first and then by name (in alphabetical order). The proposed changes in this PR were necessary to build the documentation of `python-igraph` using PyDoctor (see the results [here](https://igraph.org/python/doc/api/index.html)) -- some of igraph's classes contain quite a few members and the original unsorted list was hard to use.

### CODE REVIEW

File: pydoctor/templatewriter/pages/__init__.py
Comment (tristanlatr) [inline pydoctor/templatewriter/pages/__init__.py:284]: I think this will hide the `__init__` methods and `__init__` modules. As you can in the docs generated for this PR: [here](https://pydoctor--379.org.readthedocs.build/en/379/api/pydoctor.templatewriter.writer.TemplateWriter.html) for instance, the `__init__` method do not show up...

Let's note that we are discussing about how to display of constructors [here](https://github.com/twisted/pydoctor/issues/305). And I think the PR #360 will solve the issue with `__init__` module everywhere.

About the sorting order, @mthuurne will correct me if I'm wrong, but I think it's a design decision (probably discutable) to list members in the same order they have been defined in the code. So maybe this should be an option like `--sort-members` for instance. Also sorting by kind would be useful maybe?

**Review by tristanlatr — CHANGES_REQUESTED** [review]
Hello @ntamas ,

Thanks a lot for opening the PRs. 

I'll do a first pass of reviews :)

File: pydoctor/templatewriter/pages/__init__.py
Comment (mthuurne) [reply pydoctor/templatewriter/pages/__init__.py:284]: Once #360 lands, it will no longer be necessary to filter out `__init__` modules, since `__init__.py` will be treated as the source file for the package rather than a module within the package.

I don't know whether using the source order was a design decision or just a result of the most straightforward implementation. For most projects, source order is either pure chaos or some logic that only the developer understands, so sorting the names would benefit the reader, in my opinion.

I'll post a separate reply at the top level to discuss the sort order, since that's more of a requirements discussion than a code discussion.

Comment (mthuurne) [reply pydoctor/templatewriter/pages/__init__.py:284]: Actually, this code is inside `ClassPage`, so it would only be filtering out `__init__()` methods. If there is a reason why you wouldn't want to see the `__init__()` methods listed, please document it in #305. But for the time being, we'd lose information by simply dropping them unconditionally for all projects.

**Comment by mthuurne** [comment]
Regarding the order of members, source order is usually not the most intuitive to the reader, so I agree that we should pick a different sort order by default.

While I write "by default" here, I'd like to try and define a single fixed sort order that will be acceptable for all projects. If it turns out that is not possible, we could consider adding command line options to enable different sort orders. For the sake of maintainability, I'd rather not introduce more options unless they are really needed.

I think that we should apply sorting to modules as well, not only classes.

As @tristanlatr suggested, sorting by "kind" (class method, instance method, class variable, instance variable, property etc.) might be useful. In particular, grouping things that can be accessed on a class together and things that can be accessed on an instance. Maybe it would make more sense to split this into two separate tables though instead of using it as a sort order?

In the sidebar (#370) methods and attributes are separated and each of those categories is sorted alphabetically. Would it make sense to use the same grouping and sorting for the sidebar and the main content?

File: pydoctor/templatewriter/pages/__init__.py
Comment (ntamas) [reply pydoctor/templatewriter/pages/__init__.py:284]: This method was blindly copied from `PackagePage` where `__init__` is also filtered, although I guess that `__init__` refers to `__init__.py` there, not the `__init__` method, right? If that is the case, then indeed it should not be filtered.

**Comment by tristanlatr** [comment]
Hi,

I agree we should group the members by kind. 

I think the exhaustive list of "kind"s is something like this :  
- Class 
- Class Method
- Instance Method
- Class Variable
- Instance Variable
- Variable
- Property
- Function
- Module
- Package
- Interface ?
- Exception (TODO): Shows as Class
- Constant (TODO): Shows as Variable

Just for the record, Dash proposes filters for Twisted docs like follows: 

by default 2021-04-08 at 2 27 19 PM

As talked in #364, we should also switch to an Enum to handle the kind values, then we can safely refactor the Pages to show subsections with all kinds. 

For the short term maybe we could replace 

```
key=lambda o2:(-o2.privacyClass.value, o2.fullName()))
```

by 
```
key=lambda o2:(-o2.privacyClass.value, o2.kind, o2.fullName()))
```

And see if the output sorted by privacy, kind, then name is acceptable ?

**Comment by ntamas** [comment]
I gave it a go with the igraph documentation and it seemed perfectly fine for me; I'll push a new commit here soon.

**Comment by tristanlatr** [comment]
> it seemed perfectly fine for me

I'm unsure honestly, The `__init__` used to show up first if it was declared first. Now it's lost in the middle somewhere... 
I think the order of kins should be as follow: Packages, Modules, Classes, Class Methods,  Static Methods, (Instance) Methods, Class Variables, Instance Variables, Variables,  Functions. 


We could define something like :

```python

class KindClass(Enum):
    Package = 1000
    Module = 900
    Interface = 850
    Class = 800
    Class_Method = 700
    Static_Method = 600
    Instance_Method = 500
    Async_Class_Method = 480
    Async_Static_Method = 460
    Async_Instance_Method = 440
    Function = 400
    Async_Function = 350
    Class_Variable = 300
    Instance_Variable = 200
    Property = 150
    Variable = 100
```

I'm not sure if we need to use, orderedenums: https://docs.python.org/3/library/enum.html#orderedenum .

Then, a (huge) bit if refactoring to use the new `KindClass`, and it should do the trick. 

`kind` apprears 111 times in our codes, I don't think it's really hard to refactor. 

 It's starting to be unmaintainable anyway, so we really need this. I can open a PR for that. 

But maybe there is another solution ?

**Comment by tristanlatr** [comment]
So once #382 is merged we can use:
```
key=lambda o2:(-o2.privacyClass.value, -o2.kind, o2.fullName()))
```
And Methods should show up firsts is classes. According to the Enum values.

Then, more work will be needed to group the tables by kind like in javadoc.

**Comment by tristanlatr** [comment]
Hi @ntamas,

The PR that transform the kind values to en Enum is merged, so we'll be able to have sorting order that makes sens :)

Could you update your PR with master ?

Also could you apply changes such that the same order is used for the ModulePage and PackagePage ?

Thank you,

**Comment by ntamas** [comment]
I think this is done. Let me know if I missed anything.

**Comment by tristanlatr** [comment]
It looks like only the table rows are ordered the right way but the content items in the page (starting after the tables), do not use the same order. I think the `childlist` renderer should also be changed, I assume adding the key to the `sorted` function in `CommonPage.methods()` should do it. 

by default 2021-04-26 at 10 14 53 AM

**Comment by ntamas** [comment]
It seems like `CommonPage.children()` does not sort by name, only by privacy class and kind. `ModulePage` inherits from `CommonPage` so it inherits this behaviour. If I change the implementation of `CommonPage.children()`, that could have consequences for other `CommonPage` subclasses -- OTOH, it might actually be desired even for other subclasses as well. What's your opinion, should I override `.children()` in `ModulePage` or can I fix it in `CommonPage`?

**Comment by tristanlatr** [comment]
I think you can change the sorting order in the base class, so the children will also be sorted in the ClassPage and PackagePage.

**Comment by ntamas** [comment]
Changed it in the base class now.

**Comment by tristanlatr** [comment]
Can you also edit the `CommonPage.methods()` so the child list is also using the same order ?

thanks

**Comment by ntamas** [comment]
It's done now. Sorry, I was searching for `sorted()` to identify places where I needed to intervene, this one has slipped through.

**Comment by tristanlatr** [comment]
Can you allow me to push some commits to `igraph/pydoctor` repo ?

Thanks

**Comment by ntamas** [comment]
I thought you are allowed to push commits to my PR but in case you aren't, I have sent an invite to the repo with write access.

**Review by tristanlatr — APPROVED** [review]

**Comment by tristanlatr** [comment]
Thanks!

I think it's because you forked the project with your organization profile that you needed to give me access.

**Review by mthuurne — APPROVED** [review]
Looks good; thanks!

**Pull request merged**
